distributed/tests/test_scheduler.py

```diff
@@ -126,6 +126,93 @@ async def test_decide_worker_with_restrictions(client, s, a, b, c):
     assert x.key in a.data or x.key in b.data
 
 
+@pytest.mark.parametrize("ndeps", [0, 1, 4])
+@pytest.mark.parametrize(
+    "nthreads",
+    [
+        [("127.0.0.1", 1)] * 5,
+        [("127.0.0.1", 3), ("127.0.0.1", 2), ("127.0.0.1", 1)],
+    ],
+)
+def test_decide_worker_coschedule_order_neighbors(ndeps, nthreads):
+    @gen_cluster(
+        client=True,
+        nthreads=nthreads,
+        config={"distributed.scheduler.work-stealing": False},
+    )
+    async def test(c, s, *workers):
+        """Ensure that related tasks end up on the same node"""
+        da = pytest.importorskip("dask.array")
+        np = pytest.importorskip("numpy")
+
+        if ndeps == 0:
+            x = da.random.random((100, 100), chunks=(10, 10))
+        else:
+
+            def random(**kwargs):
+                assert len(kwargs) == ndeps
+                return np.random.random((10, 10))
+
+            trivial_deps = {f"k{i}": delayed(object()) for i in range(ndeps)}
+
+            # TODO is there a simpler (non-blockwise) way to make this sort of graph?
+            x = da.blockwise(
+                random,
+                "yx",
+                new_axes={"y": (10,) * 10, "x": (10,) * 10},
+                dtype=float,
+                **trivial_deps,
+            )
```

Comment:
What sort of graph are you trying to make?  

Thoughts on the test from my PR?  Or perhaps a variant of it that uses `np.from_array` or maybe uses Zarr?

Reply:
This is a heavily modified version of the test from your PR, but I started with that verbatim. I removed the "try both axes" logic because it felt redundant, but I can easily add that back with another `parametrize`. I also made it less hardcoded to the number of keys on a worker; I found thinking about statistics of fractions more intuitive.

The graphs look like:

`ndeps == 1`:
![graph-1](https://user-images.githubusercontent.com/3309802/123185619-94dfc880-d453-11eb-9f08-e886a76f930c.png)
`ndeps == 4`:
![graph-4](https://user-images.githubusercontent.com/3309802/123185623-95785f00-d453-11eb-9d62-517aec4df1c0.png)

I though about using `from_zarr` (since that's the case we want to simulate), but I wanted to be able to test > 1 trivial dependency. This was the easiest way I could come up with to make the graph with any number of trivial dependencies.

Reply:
> I though about using from_zarr (since that's the case we want to simulate), but I wanted to be able to test > 1 trivial dependency. This was the easiest way I could come up with to make the graph with any number of trivial dependencies.

I recommend that we include a from_zarr test anyway.

At some point in the future this test is going to break and it will be hard for people to figure out what it does.  They may xfail it.  The from_zarr test will, I think, be easier to reason about and will be very close to the workflow that we care about.  I think that if we can make something that has the full reach and intuitiveness of an integration test, but the speed and light weight of a unit test, then that would be ideal.

Reply:
> I removed the "try both axes" logic because it felt redundant

FWIW I far prefer redundancy over complexity.  I think that another challenge with using blockwise is that very few maintainers understand it well.  If we can use more common API, like `.mean(axis=)` then I think that we open the door for more people to understand and help maintain this code.  If that means that we copy-paste a little bit then I think that that trade-off is well worth it.

In general I encourage you to try to write code that someone far less sophisticated than you will find easy to understand.